### PR TITLE
Switch assets fast

### PR DESCRIPTION
**Pull request by Reizoukodesu**
Added a eyedropper tool toggle button for the editor which allows to switch quickly between tileset categories and select the clicked tile template as current brush. See #10643

### CODE REVIEW

**Comment by pchote** [comment]
Could you please try and make this work as an `IEditorBrush` implementation, instead of adding the logic to the viewport widget?

**Comment by Sapena** [comment]
I've extracted the eyedropper brush logic to a new brush type.
I also added a visual feedback for the tool, I've called it EditorHighlightLayer. This could be used for other functions as well, as mentioned in [#10652](https://github.com/OpenRA/OpenRA/issues/10652#issuecomment-177339989). Should I extract the EditorHighlightLayer functionality to a different pull request?

File: OpenRA.Mods.Common/Traits/World/EditorHighlightLayer.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		public readonly string Image = "editor-overlay";
+
+		[SequenceReference("Image")]
+		[Desc("Sequence to use for the highlight overlay.")]
+		public readonly string HighlightSequence = "highlight";
+
+		public virtual object Create(ActorInitializer init) { return new EditorHighlightLayer(init.Self, this); }
+	}
+
+	public class EditorHighlightLayer : IWorldLoaded, IPostRender
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/EditorHighlightLayer.cs]: There's a lot of duplication between this and the existing `EditorSelectionLayer`. Can you merge the two together?

File: OpenRA.Mods.Common/Widgets/EditorViewportControllerWidget.cs [hunk 2]
```diff
@@ -35,7 +36,28 @@ public EditorViewportControllerWidget(World world, WorldRenderer worldRenderer)
 			CurrentBrush = defaultBrush = new EditorDefaultBrush(this, worldRenderer);
 		}
 
+		public void ToggleEyedropping()
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/EditorViewportControllerWidget.cs]: The modifications in this file aren't needed.  Just set the brush type directly from the `MapEditorLogic` (see how the copy/paste brush does it).

File: OpenRA.Mods.Common/Widgets/Logic/Editor/TileSelectorLogic.cs [hunk 3]
```diff
@@ -50,6 +52,13 @@ public TileSelectorLogic(Widget widget, WorldRenderer worldRenderer, Ruleset mod
 			IntializeTilePreview(widget, worldRenderer, tileset, categories.First());
 		}
 
+		public void SwitchCategory(WorldRenderer worldRenderer, TileSet tileset, string category)
+		{
+			var tileCategorySelector = tileSelectorContainer.Get<DropDownButtonWidget>("TILE_CATEGORY");
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Editor/TileSelectorLogic.cs]: This can be cached from the constructor.

File: mods/ra/chrome/editor.yaml [hunk 4]
```diff
@@ -394,3 +394,9 @@ Container@EDITOR_WORLD_ROOT:
 			Align: Left
 			Font: Bold
 			Contrast: true
+		Button@EYEDROPPER_BUTTON:
+			X: 665
+			Width: 120
+			Height: 25
+			Text: Eyedropper tool
```
Comment (pchote) [inline mods/ra/chrome/editor.yaml]: This label is too long, and overlaps the minimap.  I suggest renaming this to "Select terrain" and moving the button.

File: mods/ra/rules/world.yaml [hunk 5]
```diff
@@ -174,3 +174,4 @@ EditorWorld:
 	EditorActorLayer:
 	EditorResourceLayer:
 	EditorSelectionLayer:
+	EditorHighlightLayer:
```
Comment (pchote) [inline mods/ra/rules/world.yaml]: Please add the feature to all the default mods, not just RA.

File: OpenRA.Mods.Common/Traits/World/EditorHighlightLayer.cs (hunk 1 above)
Comment (Sapena) [inline OpenRA.Mods.Common/Traits/World/EditorHighlightLayer.cs]: I couldn't find a good way to merge them together, so I extracted the shared logic to a parent class, is this ok?

**Comment by Sapena** [comment]
Hope all issues are fixed. Let me know if there's something more to do.

**Comment by Mailaender** [comment]
Can you squash the fixup commits using `git rebase`?

File: OpenRA.Mods.Common/Traits/World/EditorCellRegionLayer.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+				return;
+
+			palette = wr.Palette(Info.Palette);
+		}
+
+		public abstract void Clear();
+
+		public virtual void RenderAfterWorld(WorldRenderer wr, Actor self)
+		{
+			if (wr.World.Type != WorldType.Editor)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/EditorCellRegionLayer.cs]: You immediately return if this is false, too, so this check is a no-op.

**Comment by pchote** [comment]
Any progress here @Sapena?

**Comment by reaperrr** [comment]
There hasn't been any activity in two months, so I'm closing this for now to clear the PR queue.
@Sapena Feel free to reopen it when you have the time to rebase and adress the requests.

**Pull request closed without merging**